lib/common/cmdline.c

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <config.h>
+#include <glib.h>
+
+#include <crm/common/cmdline.h>
+#include <crm/common/util.h>
+
+GOptionContext *
+pcmk__build_arg_context(const char *param_string, const char *fmts,
+                        gboolean *version, gboolean *quiet, GOptionArgFunc verbose_cb,
+                        char **output_ty, char **output_dest) {
+    char *desc = crm_strdup_printf("Report bugs to %s\n", PACKAGE_BUGREPORT);
+    char *output_as_desc = crm_strdup_printf("Specify the format for output, one of: %s", fmts);
+    GOptionContext *context;
+
+    GOptionEntry main_entries[6] = {
+        { "version", '$', 0, G_OPTION_ARG_NONE, version,
+          "Display version information and exit.",
+          NULL },
+        { "verbose", 'V', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, verbose_cb,
+          "Increase debug output (may be specified multiple times).",
+          NULL },
+        { "quiet", 'q', 0, G_OPTION_ARG_NONE, quiet,
+          "Be less descriptive in output.",
+          NULL },
+        { "output-as", 0, 0, G_OPTION_ARG_STRING, output_ty,
+          output_as_desc,
+          "FORMAT" },
+        { "output-to", 0, 0, G_OPTION_ARG_STRING, output_dest,
+          "Specify the destination for formatted output, \"-\" for stdout or a filename", "DEST" },
+
+        { NULL }
+    };
+
+    context = g_option_context_new(param_string);
+    g_option_context_set_description(context, desc);
+    g_option_context_add_main_entries(context, main_entries, NULL);
+
+    free(desc);
+    /* output_as_desc must not be freed in this function or it won't appear in output */
```

Comment:
Hmm, I'm guessing it will never be freed (before exit)? Not a big deal in practice, but valgrind may flag it. Not worth jumping through hoops to fix.

Reply:
I should call `g_option_context_free` for starters.  That'll probably take care of it.

Reply:
While that's a good idea :) that's a const* in GOptionEntry, so it won't get freed.

I do have an idea though -- if we have that common args struct, we can keep the pointer in there to remember it, as well as a verbose counter. Then we can make the verbose_cb static here rather than passed as an arg, and have a "process common args" function that would call g_option_context_free() and free the pointer, and create and return the output object. Feel free to consider that overkill however and stick with the current approach.

Reply:
Okay that's a little bit roundabout, but I think I've got it figured out.  New version coming in a few minutes.